test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs

```diff
@@ -2544,6 +2544,98 @@ public async Task ExecuteAsync_WithDowngradesInPrunedSubgraph_DoesNotRaiseNU1605
             result.LockFile.Libraries.Count.Should().Be(4);
         }
 
+        /// <summary>
+        /// A 1.0 -> D 1.0 (Central transitive)
+        ///       -> B 1.0 -> D 3.0 (Central transitive - should be ignored)
+        ///                -> C 1.0 -> D 2.0
+        /// </summary>
+        [Fact]
+        public async Task ExecuteAsync_TransitiveDependenciesFromNonRootLibraries_AreIgnored()
```

Comment:
So I created this test with nuget.exe instead (basically, I want to avoid constructing the PackageSpec manually) and here's what I have. 

```cs
/// <summary>
        /// A 1.0 -> D 1.0 (Central transitive)
        ///       -> B 1.0 -> D 3.0 (Central transitive - should be ignored)
        ///                -> C 1.0 -> D 2.0
        /// </summary>
        [Fact]
        public async Task NuGetExeRestore_Test()
        {
            // Arrange
            using var pathContext = new SimpleTestPathContext();

            // Set up solution, project, and packages
            var solution = new SimpleTestSolutionContext(pathContext.SolutionRoot);

            var projectA = SimpleTestProjectContext.CreateNETCoreWithSDK(
                "a",
                pathContext.SolutionRoot,
                "net472");
            var projectB = SimpleTestProjectContext.CreateNETCoreWithSDK(
                "b",
                pathContext.SolutionRoot,
                "net472");
            var projectC = SimpleTestProjectContext.CreateNETCoreWithSDK(
                "c",
                pathContext.SolutionRoot,
                "net472");

            await SimpleTestPackageUtility.CreateFolderFeedV3Async(
                pathContext.PackageSource,
                PackageSaveMode.Defaultv3,
                new SimpleTestPackageContext("d", "1.0.0"),
                new SimpleTestPackageContext("d", "2.0.0"),
                new SimpleTestPackageContext("d", "3.0.0")
                );

            solution.Projects.Add(projectA);
            solution.Projects.Add(projectB);
            solution.Projects.Add(projectC);
            solution.Create(pathContext.SolutionRoot);

            AddProjectReference(projectA, projectB);
            AddProjectReference(projectB, projectC);
            AddPackageReferenceToProject(projectC);

            CreateDirectoryPackagesPropsWithVersionForPackageD(pathContext, "a", "1.0.0");
            CreateDirectoryPackagesPropsWithVersionForPackageD(pathContext, "b", "3.0.0");
            CreateDirectoryPackagesPropsWithVersionForPackageD(pathContext, "c", "2.0.0");

            var args = new string[] {
                    "restore",
                    solution.SolutionPath,
                    "-Verbosity",
                    "detailed",
                };

            // Act
            var r = CommandRunner.Run(
                Util.GetNuGetExePath(),
                pathContext.WorkingDirectory.Path,
                string.Join(" ", args),
                waitForExit: true);

            // Assert
            r.Success.Should().BeFalse(because: r.AllOutput);
        }

        private static void CreateDirectoryPackagesPropsWithVersionForPackageD(SimpleTestPathContext pathContext, string projectName, string version)
        {
            var directoryPackagesPropsContent =
                           @$"<Project>                    
                        <ItemGroup>
                            <PackageVersion Include=""D"" Version=""{version}"" />
                        </ItemGroup>
                    </Project>";
            File.WriteAllText(Path.Combine(pathContext.SolutionRoot, projectName, $"Directory.Packages.Props"), directoryPackagesPropsContent);
        }

        private static void AddProjectReference(SimpleTestProjectContext from, SimpleTestProjectContext to)
        {
            var xml = from.GetXML();

            ProjectFileUtils.AddProperty(
                xml,
                "ManagePackageVersionsCentrally",
                "true");

            ProjectFileUtils.AddItem(
                            xml,
                            "ProjectReference",
                            to.ProjectPath,
                            NuGetFramework.AnyFramework,
                            new Dictionary<string, string>(),
                            new Dictionary<string, string>());

            xml.Save(from.ProjectPath);
        }
```

This test seems to fail with your changes. 

Is there an issue with the changes or is it my test that might be missing something?

Reply:
my bad, added now.

Reply:
Your test didn't set `CentralPackageTransitivePinningEnabled` property (by default it is false).
Here is an updated version of `AddProjectReference` method for which the test doesn't fail anymore:

```
 private static void AddProjectReference(SimpleTestProjectContext from, SimpleTestProjectContext to)
        {
            var xml = from.GetXML();

            ProjectFileUtils.AddProperty(
                xml,
                "ManagePackageVersionsCentrally",
                "true");

            ProjectFileUtils.AddProperty(
                xml,
                "CentralPackageTransitivePinningEnabled",
                "true");

            ProjectFileUtils.AddItem(
                            xml,
                            "ProjectReference",
                            to.ProjectPath,
                            NuGetFramework.AnyFramework,
                            new Dictionary<string, string>(),
                            new Dictionary<string, string>());

            xml.Save(from.ProjectPath);
        }
```

Reply:
Perfect. 
Thanks for your patience with this.